docs/source/jit_type_system_1.rst

```diff
@@ -0,0 +1,628 @@
+            return True
+        return x == y
+
+    m = torch.jit.script(enum_fn)
+
+    print("Eager: ", enum_fn(Color.RED, Color.GREEN))
+    print("TorchScript: ", m(Color.RED, Color.GREEN))
+
+Example:
+
+The following example shows the case of restricted enum subclassing, where BaseColor does not define any member, thus can be subclassed by Color.
+
+.. testcode::
+
+    import torch
+    from enum import Enum
+
+    class BaseColor(Enum):
+        def foo(self):
+            pass
+
+    class Color(BaseColor):
+        RED = 1
+        GREEN = 2
+
+    def enum_fn(x: Color, y: Color) -> bool:
+        if x == Color.RED:
+            return True
+
+        return x == y
+
+    m = torch.jit.script(enum_fn)
+    print("TorchScript: ", m(Color.RED, Color.GREEN))
+    print("Eager: ", enum_fn(Color.RED, Color.GREEN))
+
+TorchScript Module Class
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+``TSModuleType`` is a special class type that is inferred from object instances created outside TorchScript. Although ``TSModuleType`` is named by the Python class of the object instance. Method ``__init__()`` of the Python class are not considered as a TorchScript method, thus it does not have to comply to TorchScript’s type checking rules.
+
+Since the type schema of module instance class is constructed directly from an instance object (created outside the scope of TorchScript), rather than inferred from ``__init__()`` like custom classes. It is possible that two objects of the same instance class type follow two different type schemas.
+
+In this sense, ``TSModuleType`` is not really a static type. Therefore, for type safety considerations, ``TSModuleType`` cannot be used in TorchScript type annotation or be composed with ``TSType``.
+
+Module Instance Class
+^^^^^^^^^^^^^^^^^^^^^
+
+TorchScript module type represents type schema of a user-defined PyTorch module instance.  When scripting a PyTorch module, the module object is always created outside TorchScript (i.e., passed in as parameter to forward()), the Python module class is treated as a module instance class so that ``__init__()`` of the Python module class are not subject to the type checking rules of TorchScript
+
+::
+
+    TSModuleType := "class" Identifier "(torch.nn.Module)" ":"
+                        ClassBodyDefinition
+
+where
+
+* ``forward()`` and other methods decorated with ``@torch.jit.export`` must be compilable to TorchScript IR and subject to TorchScript’s type checking rules
+
+Unlike custom classes, only the forward method and other methods decorated with ``@torch.jit.export``  of the module type need to be compilable to TorchScript IR. Most notably, ``__init__()`` is not considered a TorchScript method. Thus module type constructors should not be invoked within the scope of TorchScript. Instead, TorchScript module objects are always constructed outside and passed into ``torch.jit.script(ModuleObj)``.
+
+Example:
+
+.. testcode::
+
+    import torch
+    class TestModule(torch.nn.Module):
```

Comment:
I found this example (of mine :) unnecessarily complex to explain the ModuleInstanceClass. How about using this simplified example instead?

```
mport torch
  
class TestModule(torch.nn.Module):
    def __init__(self, v):
        super().__init__()
        self.x = v

    def forward(self, inc: int):
        return self.x + inc

m = torch.jit.script(TestModule(1))
print(f"First instance: {m(3)}")

m = torch.jit.script(TestModule(torch.ones([5])))
print(f"Second instance: {m(3)}")
```
And the outputs are 
```
First instance: 4
Second instance: tensor([4., 4., 4., 4., 4.])
```